### PR TITLE
add micro batches

### PR DESCRIPTION
**Pull request by lukas-blecher**
Proposed in issue #147

### CODE REVIEW

**Comment by lukas-blecher** [comment]
Maybe scale all gradients by `args.micro_batchsize/args.batchsize` because right now the gradients are summed over for each mirco batch, resulting in a larger gradient norm on average.
But I've tested it out on a toy model and this constant factor did not hinder convergence.
This can also be compensated by choosing other betas and initial learning rate in case of adam optimizer.

Still might be better for consistency.
Add 
```python
for p in model.parameters():
    p.grad=p.grad*microbatch/args.batchsize
```
before https://github.com/lukas-blecher/LaTeX-OCR/blob/4c94f3da2a41e8991378be19e8c8e7c56c478b51/pix2tex/train.py#L57

**Comment by TITC** [comment]
As always, learned from your code and comment.

I found some code on some [websites](https://gist.github.com/thomwolf/ac7a7da6b1888c2eeac8ac8b9b05d3d3) and they do not directly average gradient but average the loss before backward propagation.


I think they come up to same result. Although it named gradient descent but in fact it's [Directional derivative](https://en.wikipedia.org/wiki/Directional_derivative). Because the direction is fixed when the architecture is determined.

image

I not sure if this is the correct understanding, I want listening your opinion. @lukas-blecher

**Comment by lukas-blecher** [comment]
Yes, I think you're right it is equivalent.
Scaling the loss would be computationally more efficient
